### ISSUE TITLE
Enum.compareByIndex in Dart 2.15.0-268-18.beta performs a difference and not a comparison

### ISSUE THREAD

**Issue by giorgiofran** [dart-lang/language#1993]
If you run the following:
```dart
enum T5 { zero, one, two }

void main() {
  print(Enum.compareByIndex<T5>(T5.two, T5.zero));
}
```
The result is `2` while I was expecting `1`.

Giving a look to the source I see this:
```dart
@Since("2.15")
  static int compareByIndex<T extends Enum>(T value1, T value2) =>
      value1.index - value2.index;
```
So, effectively, the method performs a difference, while the description suggest that it should be a comparison:
```
value1.index.compareTo(value2.index);
```

Either the description or the statement are wrong.

**Comment by giorgiofran** [comment]
Looking at Comparable specs, it seems that this is not a bug, because any positive value is a valid return value not only one.
Nevertheless, it would be more consistent with `int` comparison if one was returned. I suggest to change this method like I exposed above.

**Comment by Levi-Lesches** [comment]
The documentation says: 

> ```dart
>  /// Compares two enum values by their [index].
>  ///
>  /// A generic [Comparator] function for enum types which
>  /// orders enum values by their [index] value, which corresponds
>  /// to the source order of the enum element declarations in
>  /// the `enum` declaration.
> ```

And the docs for `Comparator`: 
> A Comparator function represents such a total ordering by returning
> - a negative integer if a is smaller than b,
> - zero if a is equal to b, and
> - a positive integer if a is greater than b.

Seems like the function is working perfectly as intended. The docs make it pretty clear that you're not supposed to care about the specific value returned by the compare function, rather just whether it's positive, negative, or zero.

**Comment by lrhn** [comment]
Yes, the implementation is correct and deliberate. Since we know that the values are small non-negative integers, the subtraction should be correct and more efficient than, basically, doing `(value1.index - value2.index).sign` and normalizing the values to {-1, 0, +1}. We know that subtracting two non-negative integers cannot overflow.

For full integers, we do not know that, and the subtraction could potentially overflow, which is why we do something more complicated.

Always use `< 0`/`== 0`/`> 0` for checking the result of comparators, it's not guaranteed (and not true) that they always return zero or plus/minus one. That's just convenient values to use when the result is found by comparing non-integers.
